signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/* This module handles verification of an X509 certificate chain and Mozilla's
+ * content signature algorithm which is used for signing and authentication of
+ * updates.
+ *
+ * This provides methods to parse an X509 certificate chain out of a PEM file,
+ * perform certificate chain validation to ensure the code signing certificate
+ * can be trusted, and verifies a content signature against the code signing
+ * certificate.
+ *
+ * The content signature itself is a base64-encoded ECDSA signature, using the
+ * content prefixed with "Content-Signature\x00" as the input data.
+ */
+
+use hex;
+use asn1_rs::ToDer;
+use ring::digest;
+use x509_parser::prelude::*;
+
+use oid_registry::{
+    OID_SIG_ECDSA_WITH_SHA256,
+    OID_SIG_ECDSA_WITH_SHA384,
+};
+
+pub struct Balrog<'a> { 
+    pub chain: Vec<X509Certificate<'a>>,
+    pub root_hash: Vec<u8>
+}
+
+// Errors that can be returned from the Balrog module.
+#[derive(Clone, Debug, PartialEq, thiserror::Error)]
+pub enum BalrogError {
+    #[error("Chain subject/issuer mismatch")]
+    ChainSubjectMismatch,
+    #[error("Certificate expired")]
+    CertificateExpired,
+    #[error("Issuer unauthorized for certificate signing")]
+    IssuerUnauthorized,
+    #[error("Certificate unauthorized for code signing")]
+    CodeSignUnauthorized,
+    #[error("Root hash parse failed")]
+    RootHashParseFailed,
+    #[error("Root hash mismatch")]
+    RootHashMismatch,
+    #[error("Hostname mismatch")]
+    HostnameMismatch,
+    #[error("Signature decoding failed")]
+    SignatureDecodeError,
+    #[error("PEM decoding error")]
```

Comment:
Should this be #[error("PEM decoding error: {0:?}")] ?

Reply:
Yes, thank you!